File: src/main/java/com/hp/jipp/encoding/LangStringType.java
package com.hp.jipp.encoding;

import com.google.common.base.Optional;
import com.google.common.collect.Lists;
import com.hp.jipp.util.BuildError;
import com.hp.jipp.util.Util;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.List;

/** An language-encoded string attribute type */
public class LangStringType extends AttributeType<LangString> {
    private static final String TYPE_NAME = "LangString";

    static final Attribute.Encoder<LangString> ENCODER = new Attribute.Encoder<LangString>(TYPE_NAME) {
        @Override
        public LangString readValue(DataInputStream in, Tag valueTag) throws IOException {
            byte[] bytes = OctetStringType.ENCODER.readValue(in, valueTag);
            DataInputStream inBytes = new DataInputStream(new ByteArrayInputStream(bytes));

            String lang = new String(Attribute.readValueBytes(inBytes), Util.UTF8);
            String string = new String(Attribute.readValueBytes(inBytes), Util.UTF8);
            return LangString.of(string, lang);
        }

        @Override
        public void writeValue(DataOutputStream out, LangString value) throws IOException {
            ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
            DataOutputStream dataOut = new DataOutputStream(bytesOut);
            Optional<String> lang = value.getLang();
            if (!lang.isPresent()) {
                throw new BuildError("Cannot write a LangString without a language");
            }
            writeValueBytes(dataOut, lang.get().getBytes(Util.UTF8));
            writeValueBytes(dataOut, value.getString().getBytes(Util.UTF8));
            OctetStringType.ENCODER.writeValue(out, bytesOut.toByteArray());
        }

        @Override
        public boolean valid(Tag valueTag) {
            return valueTag == Tag.NameWithLanguage || valueTag == Tag.TextWithLanguage;
        }
    };

    public LangStringType(Tag tag, String name) {
        super(ENCODER, tag, name);
    }

    @Override
    @SuppressWarnings({"PMD.UselessParentheses", "unchecked"})
    public Optional<Attribute<LangString>> of(Attribute<?> attribute) {
        if (!(attribute.getValueTag().equals(Tag.NameWithoutLanguage) && getTag().equals(Tag.NameWithLanguage)) ||
                (attribute.getValueTag().equals(Tag.TextWithoutLanguage) && getTag().equals(Tag.TextWithLanguage))) {
            return Optional.absent();
        }
        // TODO: If we don't know the language this is actually a dangerous thing to do
        // Apply conversion from StringType to a LangStringType on demand
        return Optional.of(of(Lists.transform((List<String>) attribute.getValues(), LangString.FromStringFunc)));
    }
}

File: src/main/java/com/hp/jipp/encoding/LangStringType.kt
package com.hp.jipp.encoding

import com.hp.jipp.util.BuildError

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** An language-encoded string attribute type  */
class LangStringType(tag: Tag, name: String) : AttributeType<LangString>(LangStringType.ENCODER, tag, name) {

    /** Return an [Attribute] of this type */
    override fun of(attribute: Attribute<*>): Attribute<LangString>? {
        if (!validInputTag(attribute.valueTag)) {
            return null
        }

        // TODO: If we don't know the language this is actually a dangerous thing to do
        // Apply conversion from StringType to a LangStringType on demand
        return of(attribute.values.map { LangString(it as String) })
    }

    private fun validInputTag(fromTag: Tag): Boolean =
            when (fromTag) {
                Tag.nameWithoutLanguage -> tag == Tag.nameWithLanguage
                Tag.textWithoutLanguage -> tag == Tag.textWithLanguage
                else -> fromTag == tag
            }

    companion object {
        private val TYPE_NAME = "LangString"

        @JvmField
        val ENCODER: SimpleEncoder<LangString> = object : SimpleEncoder<LangString>(TYPE_NAME) {
            @Throws(IOException::class)
            override fun readValue(input: DataInputStream, valueTag: Tag): LangString {
                val bytes = OctetStringType.ENCODER.readValue(input, valueTag)
                val inBytes = DataInputStream(ByteArrayInputStream(bytes))

                val lang = inBytes.readString()
                val string = inBytes.readString()
                return LangString(string, lang)
            }

            @Throws(IOException::class)
            override fun writeValue(out: DataOutputStream, value: LangString) {
                val bytesOut = ByteArrayOutputStream()
                val dataOut = DataOutputStream(bytesOut)
                value.lang ?: throw BuildError("Cannot write a LangString without a language")
                dataOut.writeString(value.lang)
                dataOut.writeString(value.string)
                OctetStringType.ENCODER.writeValue(out, bytesOut.toByteArray())
            }

            override fun valid(valueTag: Tag): Boolean {
                return valueTag === Tag.nameWithLanguage || valueTag === Tag.textWithLanguage
            }
        }
    }
}
